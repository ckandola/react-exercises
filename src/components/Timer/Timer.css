:root {
    --accent:rgb(0, 57, 80);
    --textColor:rgb(89, 173, 206);
    --btn: rgb(59, 59, 168);
    --btn-padding: 0.3rem;
}

.timer-div {
    color:var(--textColor);
    text-align: center;
    background: rgb(69, 69, 136);
    border-radius: 0.75rem;
    width: 80%;
    margin:auto;
    padding-bottom: 0.5rem;
}

.timer-div h2 {
    margin: 0.5rem auto;
    padding: 0.5rem;
    text-shadow: 1px 3px 3px var(--accent);
}

.timer-minbtn {
    width: 8rem;
    margin: 0rem;
}

.timer-minbtn-span {
    display: block;
    margin:auto;
    width:50%;
}

.timer-statusbtn {
    margin: 0.5rem;
    padding: var(--btn-padding);
    border-radius: 0.25rem;
    box-shadow: 0px 2px 2px black;
}

.timer-btn {
    text-transform: uppercase;
    background: var(--btn);
    color:white;
    font-weight: bold;
    letter-spacing: 0.1rem;
}

.timer-form {
    margin:0.5rem;
    font-size: 1rem;
}

.timer-input {
    border-radius: 0.5rem;
    padding-left: 0.5rem;
    margin-right:0.5rem;
}

.timer-submit {
    padding: 0px var(--btn-padding);
    border-radius: 0.25rem;
    box-shadow: 0px 2px 2px black;
}

/* when screen is large... */
@media screen and (min-width: 1100px) {
    .timer-div {
        width: 40%;
    }

    .timer-minbtn {
        margin: 0.5rem;
    }

    .timer-minbtn-span {
        display:flex;
        width:max-content;
    }
}
