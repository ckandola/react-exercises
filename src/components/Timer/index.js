import React, {useState, useEffect, useCallback} from 'react';
import soundfile from './level_up.m4r';
import './Timer.css';

const Timer = () => {
    const initialText = 'Enter countdown value';
    const noCountdown = 'No time remaining';
    const [inputText, setInputText] = useState(initialText);
    const [remainingTime, setRemainingTime] = useState(0);
    const [canResume, setCanResume] = useState(false);
    const [choice, setChoice] = useState();
    const [countdownText, setCountdownText] = useState(noCountdown);
    const timeChoices = [1, 5, 10, 15];

    const playAudio = () => {
        const sound = new Audio(soundfile);
        sound.play();
    }

    const isNumber = entry => {
        return /^[0-9]*$/.test(entry);
    }

    const cancel = () => {
        setRemainingTime(0);
        setCanResume(false);
        setCountdownText(noCountdown);
    }

    const reset = () => {
        setRemainingTime(choice);
        setCanResume(true);
    }

    
    const refineCountdown = useCallback(() => {
        let text = 'Time Left: ';
        if(remainingTime < 60) {
            text += `${remainingTime} second${remainingTime > 1 ? 's' : ''}`;
            setCountdownText(text);
        } else if(remainingTime < 3600) {
            const minutes = parseInt(remainingTime / 60);
            const seconds = parseInt(remainingTime % 60);
            text += ` ${minutes} minute${minutes > 1 ? 's' : ''}`;
            text += seconds > 0 ? `, ${seconds} second${seconds !== 1 ? 's' : ''}` : '';
            setCountdownText(text);
        } else if(remainingTime < 86400) {
            const mod3600 = parseInt(remainingTime % 3600);
            const hours = parseInt(remainingTime / 3600);
            const minutes = parseInt(mod3600 / 60);
            const seconds = parseInt(mod3600 % 60);
            text += `${hours} hour${hours !== 1 ? 's' : ''}`;
            text += minutes > 0 ? `, ${minutes} minute${minutes !== 1 ? 's' : ''}` : '';
            text += seconds > 0 ? `, ${seconds} second${seconds !== 1 ? 's' : ''}` : '';
            setCountdownText(text);
        } else {
            setCountdownText('this will take a while')
        }
    }, [remainingTime]);

    useEffect(() => {
        if(canResume) {
            if(remainingTime > 0){
                refineCountdown();
                const timer = setTimeout(() => {
                    setRemainingTime(remainingTime - 1);
                }, 1000)
                return () => clearTimeout(timer);
            } else {
                setCountdownText(noCountdown);
                playAudio();
            }
        }
    }, [remainingTime, canResume, refineCountdown])

    const handleClick = countdown => {
        setRemainingTime(countdown);
        setCanResume(true);
        setChoice(countdown);
    }

    const handleTextChange = entry => {
        setInputText(entry.target.value);
    }

    const handleSubmit = () => {
        if(isNumber(inputText)){
            const num = parseInt(inputText);
            setRemainingTime(num);
            setCanResume(true);
            setChoice(num);
        }
    }

    return (
        <div className="timer-div">
            <h2>Timer</h2>
            <span className="timer-minbtn-span">
                {timeChoices.map(choice => {
                    return (<button key={choice} className="timer-btn timer-minbtn" onClick={() => {handleClick(choice)}}>{choice} seconds</button>);
                })}
            </span>
            <h2>{countdownText}</h2>
            <span>
                <button className="timer-btn timer-statusbtn" onClick={() => setCanResume(!canResume)} 
                        disabled={remainingTime === 0}>Pause/Resume</button>
                <button className="timer-btn timer-statusbtn" onClick={cancel}>Cancel</button>
                <button className="timer-btn timer-statusbtn" onClick={reset}>Reset</button>
            </span>
            <div className="timer-form">
                <input className="timer-input" value={inputText} 
                    onFocus={() => {if(!isNumber(inputText)) setInputText('')}} 
                    onChange={handleTextChange}
                    onBlur={() => setInputText(inputText ? inputText : initialText)}/>
                <input type="submit" className="timer-btn timer-submit" onClick={handleSubmit}/>
            </div>
        </div>
    )
}

export default Timer;