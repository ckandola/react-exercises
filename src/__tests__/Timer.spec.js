import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import {configure, shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import Timer from '../components/Timer';

configure({adapter: new Adapter()});

describe('Timer tests', () => {
    it('matches the snapshot', () => {
        const wrapper = shallow(<Timer />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    // Does not simulate click, so does not work/advance timer
    // TODO: figure out why this fails
    it.skip('waits 5 seconds when the 5 second button is pressed', () => {
        jest.useFakeTimers();
        const wrapper = shallow(<Timer />);
        expect(wrapper.find('h2').at(1).getElement().props.children).toEqual('No time remaining');
        wrapper.find('.timer-minbtn').at(1).simulate('click');
        expect(wrapper.find('h2').at(1).getElement().props.children).toEqual('Time Left: 5 seconds');
        jest.advanceTimersByTime(5000);
        expect(wrapper.find('h2').at(1).getElement().props.children).toEqual('No time remaining');     
        jest.useRealTimers();   
    });

    it.skip('pauses countdown when Pause/Resume is pressed after countdown has started', () => {

    });

    it.skip('resumes countdown when Pause/Resume is pressed after paused', () => {

    });

    it.skip('resets last countdown value when Reset is pressed after countdown has started', () => {

    });

    it.skip('cancels countdown when Cancel is pressed after countdown has started', () => {

    });


});
