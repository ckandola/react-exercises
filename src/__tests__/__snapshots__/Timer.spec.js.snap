// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer tests matches the snapshot 1`] = `
<div
  className="timer-div"
>
  <h2>
    Timer
  </h2>
  <span
    className="timer-minbtn-span"
  >
    <button
      className="timer-btn timer-minbtn"
      key="1"
      onClick={[Function]}
    >
      1
       seconds
    </button>
    <button
      className="timer-btn timer-minbtn"
      key="5"
      onClick={[Function]}
    >
      5
       seconds
    </button>
    <button
      className="timer-btn timer-minbtn"
      key="10"
      onClick={[Function]}
    >
      10
       seconds
    </button>
    <button
      className="timer-btn timer-minbtn"
      key="15"
      onClick={[Function]}
    >
      15
       seconds
    </button>
  </span>
  <h2>
    No time remaining
  </h2>
  <span>
    <button
      className="timer-btn timer-statusbtn"
      disabled={true}
      onClick={[Function]}
    >
      Pause/Resume
    </button>
    <button
      className="timer-btn timer-statusbtn"
      onClick={[Function]}
    >
      Cancel
    </button>
    <button
      className="timer-btn timer-statusbtn"
      onClick={[Function]}
    >
      Reset
    </button>
  </span>
  <div
    className="timer-form"
  >
    <input
      className="timer-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="Enter countdown value"
    />
    <input
      className="timer-btn timer-submit"
      onClick={[Function]}
      type="submit"
    />
  </div>
</div>
`;
